File: compiler/bench/binarysearch.fs
\ expected: 499500
\ Binary search benchmark - search sorted array, 100 iterations

1000 constant SIZE
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: init-arr ( -- )
  SIZE 0 do i i arr! loop ;

: binary-search ( val -- idx|-1 )
  0 SIZE 1-
  begin 2dup <= while
    2dup + 2/
    dup arr@ 4 pick = if
      nip nip nip exit
    then
    dup arr@ 4 pick < if
      nip 1+
    else
      rot drop swap 1-
    then
  repeat
  2drop drop -1 ;

: bench-search ( -- sum )
  0
  SIZE 0 do
    i binary-search +
  loop ;

: main ( -- )
  init-arr
  0
  100 0 do
    drop bench-search
  loop
  . cr ;

File: compiler/bench/binarysearch.c
// expected: 499500
// Binary search benchmark - search sorted array, 100 iterations

#include <stdio.h>

#define SIZE 1000

int arr[SIZE];

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = i;
    }
}

int binary_search(int val) {
    int lo = 0, hi = SIZE - 1;
    while (lo <= hi) {
        int mid = (lo + hi) / 2;
        if (arr[mid] == val) {
            return mid;
        } else if (arr[mid] < val) {
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    return -1;
}

long bench_search(void) {
    long sum = 0;
    for (int i = 0; i < SIZE; i++) {
        sum += binary_search(i);
    }
    return sum;
}

int main(void) {
    init_arr();
    long result = 0;
    for (int i = 0; i < 100; i++) {
        result = bench_search();
    }
    printf("%ld\n", result);
    return 0;
}
